From ba92a11cf3e5c2989343ce02a427d061917c01f0 Mon Sep 17 00:00:00 2001
From: vpalvai <vpalvai@codeaurora.org>
Date: Mon, 2 Apr 2018 13:53:00 -0700
Subject: [PATCH] ComputeLibrary: add cmakelist and install directives

 - add CMakelist file and install directives to compile arm
   compute library

Change-Id: I18e95035b59472b7cd4ec5bbd92d8db519c82ff4
---
 CMakeLists.txt | 94 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 94 insertions(+)
 create mode 100644 CMakeLists.txt

diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..dfadb24
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,94 @@
+cmake_minimum_required(VERSION 2.8)
+
+# set target name
+SET(TARGET_NAME arm_compute)
+
+option( ACL_NEON OFF )
+option( ACL_OPENCL OFF )
+
+set( ACL_NEON ON )
+set( ACL_OPENCL ON )
+
+FILE(GLOB core_files
+     "src/core/*.cpp"
+     "src/core/CPP/*.cpp"
+     "src/core/CPP/kernels/*.cpp"
+     "src/runtime/CPP/CPPScheduler.cpp"
+     "src/runtime/CPP/SingleThreadScheduler.cpp")
+
+FILE(GLOB runtime_files "src/runtime/*.cpp")
+
+INCLUDE_DIRECTORIES(BEFORE ${CMAKE_CURRENT_SOURCE_DIR} include)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -D_GLIBCXX_USE_NANOSLEEP")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DARCH_ARM")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pedantic")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fpermissive")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+#set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -march=armv7-a -mthumb")
+#set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -mfloat-abi=softfp -mfpu=neon")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wno-deprecated-declarations")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wextra -Wno-unused-parameter")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Winit-self -Wmissing-include-dirs")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-strict-overflow -Wswitch-default")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-vla -Woverloaded-virtual")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wctor-dtor-privacy -Wsign-promo")
+#set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Weffc++")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-ignored-qualifiers")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-overlength-strings -Wlogical-op")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wnoexcept -Wstrict-null-sentinel")
+
+if (CMAKE_BUILD_TYPE EQUAL Debug)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O0 -g -gdwarf-2")
+else(CMAKE_BUILD_TYPE EQUAL Debug)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3 -ftree-vectorize")
+endif (CMAKE_BUILD_TYPE EQUAL Debug)
+
+if (ACL_OPENCL)
+    FILE(GLOB core_cl_files
+         "src/core/CL/*.cpp"
+         "src/core/CL/kernels/*.cpp")
+
+    FILE(GLOB runtime_cl_files
+         "src/runtime/CL/*.cpp"
+         "src/runtime/CL/functions/*.cpp")
+    LIST(APPEND core_files ${core_cl_files})
+    LIST(APPEND runtime_files ${runtime_cl_files})
+endif (ACL_OPENCL)
+
+if (ACL_NEON)
+    FILE(GLOB core_neon_files
+         "src/core/NEON/*.cpp"
+         "src/core/NEON/kernels/*.cpp"
+         "src/core/NEON/kernels/arm64/*.cpp")
+
+    FILE(GLOB runtime_neon_files
+         "src/runtime/NEON/*.cpp"
+         "src/runtime/NEON/functions/*.cpp")
+
+    LIST(APPEND core_files ${core_neon_files})
+    LIST(APPEND runtime_files ${runtime_neon_files})
+endif (ACL_NEON)
+
+# add a target named ${TARGET_NAME}
+ADD_LIBRARY (${TARGET_NAME} SHARED ${core_files} ${runtime_files})
+INSTALL (TARGETS ${TARGET_NAME} DESTINATION lib64)
+INSTALL (DIRECTORY arm_compute DESTINATION include)
+INSTALL (DIRECTORY include/half DESTINATION include)
+INSTALL (DIRECTORY support DESTINATION include)
+
+# Generate build version info for utils.cpp file
+FILE(WRITE src/core/arm_compute_version.embed
+        "\"version-info"\")
+
+FILE(WRITE src/runtime/arm_compute_version.embed
+           "\"version-info"\")
+
+add_definitions( -DARM_COMPUTE_CL )
+add_executable (neon_convolution examples/neon_convolution.cpp utils/Utils.cpp)
+target_link_libraries(neon_convolution OpenCL arm_compute)
+set_target_properties(neon_convolution PROPERTIES LINK_FLAGS "-lpthread")
+
+add_executable (cl_convolution examples/cl_convolution.cpp utils/Utils.cpp)
+target_link_libraries(cl_convolution OpenCL arm_compute)
+set_target_properties(cl_convolution PROPERTIES LINK_FLAGS "-lpthread")
